src/nvapi_gpu.cpp

```diff
@@ -260,6 +260,108 @@ extern "C" {
         }
     }
 
+    NvAPI_Status __cdecl NvAPI_GPU_GetCurrentPstate(NvPhysicalGpuHandle hPhysicalGpu, NV_GPU_PERF_PSTATE_ID *pCurrentPstate) {
+        constexpr auto n = "NvAPI_GPU_GetCurrentPstate";
+        static bool alreadyLoggedNoNvml = false;
+        static bool alreadyLoggedHandleInvalidated = false;
+        static bool alreadyLoggedOk = false;
+
+        if (nvapiAdapterRegistry == nullptr)
+            return ApiNotInitialized(n);
+
+        if (pCurrentPstate == nullptr)
+            return InvalidArgument(n);
+
+        auto adapter = reinterpret_cast<NvapiAdapter*>(hPhysicalGpu);
+        if (!nvapiAdapterRegistry->IsAdapter(adapter))
+            return ExpectedPhysicalGpuHandle(n);
+
+        if (!adapter->HasNvml())
+            return NoImplementation(n, alreadyLoggedNoNvml);
+
+        if (!adapter->HasNvmlDevice())
+            return HandleInvalidated(str::format(n, ": NVML available but current adapter is not NVML compatible"), alreadyLoggedHandleInvalidated);
+
+        nvmlPstates_t pState;
+        auto result = adapter->GetNvmlPerformanceState(&pState);
+        *pCurrentPstate = static_cast<NV_GPU_PERF_PSTATE_ID>(result);
```

Comment:
You are casting the NVML result code (Success, failed etc) into the current perf state, this probably explains why you are only seeing pstate 0 ;)
You need to check that result is sucessful  like below and then cast `pState` to `pCurrentPstate`.